README.md
# README

## What is this?

The Python script in this repository detects when meeting participants in Microsoft Teams meetings raise or lower their virtual hand - and takes action to give it more emphasis in the real world. Microsoft did not seem to provide an application programming interface (API) ([Microsoft Q&A: Is there a way to detect raised hand during an online meeting using API](https://docs.microsoft.com/en-us/answers/questions/173770/is-there-a-way-to-detect-raised-hand-during-an-onl.html)), so this project has been started as a workaround. Related: [Microsoft Support: Raise your hand in a Teams meeting / See who has their hand raised](https://support.microsoft.com/en-us/office/raise-your-hand-in-a-teams-meeting-bb2dd8e1-e6bd-43a6-85cf-30822667b372).

There are currently two versions of the Python script: a command-line based one (`cli.py`) and a GUI-based one (`gui.py`). In both cases the detection result is output (to the console via standard output or as a text label in the GUI). Additionally, a string can be sent to a serial device when a hand is raised or lowered. Currently only two state transitions cause those events: *no hand → first hand* and *at least one hand → no hand* (e.g. not also one hand → two hands).

The device being connected to the serial can perform any reaction, e.g. play an animation, make a sound or move a stepper motor - but basically everything is possible. In my case I've used a micro:bit which does all three reactions.

It should be easy to connect different hardware and to implement own actions/reactions to raised/lowered hands (the edge detection is already in use).

It's originally meant for meetings that are "hybrid" - some people take part in presence, some only digitally/online (better wording is welcome; native speakers are welcome to suggest something else). This should allow online participants to become more visible by a group of participants in real-life.

## What isn't this?

This is not about gesture recognition - the script does not detect raised hands from webcam video streams.

## How does it work?

The detection logic is all done in the Python scripts `cli.py`/`gui.py`.

The detection is based on computer vision (OpenCV) pattern matching using a reference symbol. The script regularly takes screenshots and watches out for the raised hand symbol. That also means that the computer where the script is run needs to be connected to the meeting using Microsoft Teams and the list of participants needs to be visible. *(Depending on your operating system you may have to grant the script the rights to take screenshots - this could be hidden in application privacy settings.)*

When a participant raises his or her hand, the script should find out that the symbol is visible on the screen - depending on a specific, configurable threshold value. On change of detection state (hand/ no hand), the script writes a message to the serial interface.

In my case I've used a [BBC micro:bit](https://microbit.org/) to receive and react on the serial data. There is a pre-compiled `.hex` file contained in this repository which can be flashed on a micro:bit v2 (not tested for micro:bit v1 yet).

Depending on the configured serial message that is sent from the PC to the micro:bit, the micro:bit will (in this order)

- play an animation and/or
- move the servo montor and/or
- make a sound

when the first hand is raised or the last hand has been lowered. The source can be accessed under https://makecode.microbit.org/_LpmHpi9oMhXe.

![Blocks shown in Microsoft MakeCode for micro:bit](microbit-ide.png)

![Prototype with lowered hand](prototype_hand_lowered.jpg)



## How can I use it?

There are currently two methods to use this software. One ist to run it using a Python interpreter (recommended at this point in time). The other is to use a [pre-built release](https://github.com/maehw/ms-teams-hand-detector/releases) - executables for Windows, Linux and MacOS are available.

### Prerequisites and dependencies

First of all you need a computer running Microsoft Teams.

Take a screenshot of the raised hand symbol that is shown in Microsoft Teams in the participants list on the right hand side. Then cut it (about 32 x 32 pixels) and place it in the same folder as the Python script. Replace the dummy image file `the-hand.png` (or specify a different name). It is not part of this repository due to copyright uncertainties. Also it may be subject to change in future releases of Microsoft Teams.

You also need Python 3 and some more Python modules (when you do not use the pre-built standalone executables):

Some Python modules are required for taking screenshots and using computer vision to detect the hand (see the `import` statements at the top of the Python script):

```
pip install -r requirements.txt
```

*(Note: `opencv-python` currently needs to be downgraded when creating an executable with PyInstaller; related: https://github.com/maehw/ms-teams-hand-detector/issues/1)*

If you want to react, you may also want to use a BBC micro:bit, but any other hardware should do the trick (e.g. any  Arduino board).


### GUI version

Call the script `gui.py` using Python 3:

```
python3 gui.py
```

Every required setting can be made from within the GUI.


![GUI running the detection loop](gui-running.png)


### CLI version

Call the script `cli.py` using Python 3:

The following parameter is required if you want to send messages via serial interface (you can leave it, if output to the console is enough for your use case):

* `-p`: the port of the virtual serial import, e.g. `COM5` or `/dev/cu.usbmodem23542`

Optional arguments are:

* `-t`: the threshold used in the computer vision pattern matching for detecting the hand (the default value of `4.2e6` did the trick for me, but you may add some `print` log-debugging to find the threshold value that suits you)
* `-m`: the monitor number (defaults to `0`)
* `-r`: the name of the image resource (defaults to `the-hand.png`)


![Screenshot of the script's standard output in a terminal](executed-script.png)

#### Example calls

Only output to console (only if `debug = True`):

```
python3 cli.py
```

Output to console and send messages via serial:

```
python3 cli.py -p /dev/cu.usbmodem14202
```

Additionally set a custom threshold value:

```
python3 cli.py -p COM5 -t 23.5e4
```

Additionally select another monitor/display and define the hand symbol image resource:

```
python3 cli.py -p COM5 -t 4.2e6 -m 1 -r the-hand-modified.png
```

## How can I help improving this?

Help me testing this with your specific setup. I am happy for any feedback. There may also be some open issues or ideas - feel free to implement them and contribute.

### Contributing

Feel free to contribute. ;) Any issues and merge requests are welcome.

It would be great if you could run `check_code.py` on your code before you send a merge request.


CHANGELOG.md
# Changelog

## v0.2.0 (2021-11-15)

* Changed name of CLI version to `cli.py`
* Added new GUI version called `gui.py`

## v0.1.0 (2021-11-12)

Initial release


cli.py
import argparse
import cv2
import numpy as np
import time
import mss
import serial
from datetime import datetime

if __name__ == "__main__":
    serial_tx_errors = 0

    # configure your messages to be sent on a
    # - first hand detected and
    # - last hand lowered events,
    # where the following symbols are used:
    # - 'P' to PLAY the hand raised sound,
    # - 'S' to move the SERVO motor to the hand raised position,
    # - 'A' to play the hand raised ANIMATION,
    # - 'p' to PLAY the hand lowered sound,
    # - 's' to move the SERVO motor to the hand lowered position,
    # - 'a' to play the hand lowered ANIMATION
    serial_msg_on_raise = b'PAS\x0d'
    serial_msg_on_lowered = b'pas\x0d'
    serial_msg_on_init = serial_msg_on_lowered

    debug = True  # set to false to have a very quiet mode

    parser = argparse.ArgumentParser(description='%(prog)s')

    parser.add_argument('-p', action="store", default=None, required=False,
                        dest='serport',
                        help="serial device port name (e.g. '/dev/cu.usbmodem14202' or 'COM1')")

    parser.add_argument('-t', action="store", default=4.2e6, required=False,
                        dest='threshold',
                        help="threshold value for detection of the hand symbol (e.g. '4.2e6')")

    parser.add_argument('-m', action="store", default=0, required=False,
                        dest='monitor_number',
                        help="monitor number (e.g. '0')")

    parser.add_argument('-r', action="store", default="the-hand.png", required=False,
                        dest='resource_path',
                        help="resource path to hand image (e.g. 'the-hand.png')")

    args = parser.parse_args()
    args.threshold = float(args.threshold)
    args.monitor_number = int(args.monitor_number)

    if debug:
        print(f"serport = {args.serport}")
        print(f"threshold = {args.threshold}")
        print(f"monitor_number = {args.monitor_number}")
        print(f"resource_path = {args.resource_path}")

    serial_baudrate = 115200

    ser = args.serport
    if ser:
        ser = serial.Serial(ser, serial_baudrate, timeout=3)

    if debug:
        print(f"ser = {ser}")

    # try to send initialization message via serial
    if ser:
        ser.write(serial_msg_on_init)

    template = cv2.imread(args.resource_path)
    template = cv2.cvtColor(template, cv2.COLOR_BGR2GRAY)
    template = cv2.Canny(template, 50, 200)
    w, h = template.shape[::-1]

    last_detection = False

    with mss.mss(display=':0') as sct:
        monitors = sct.monitors

        # list monitors
        if debug:
            mon_num = 0
            for mon in monitors:
                print(f"monitor #{mon_num} = {mon}")
                mon_num = mon_num + 1

        while True:
            last_time = time.time()

            img = sct.grab(monitors[args.monitor_number])
            img = np.array(img)
            gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
            edged = cv2.Canny(gray, 50, 200)

            result = cv2.matchTemplate(edged, template, cv2.TM_CCOEFF)
            (_, detection_max_val, _, detection_max_loc) = cv2.minMaxLoc(result)

            loc = np.where(result >= args.threshold)
            detection = len(loc[0]) > 0

            if debug:
                now = datetime.now()
                current_time = now.strftime("%H:%M:%S")

                fps = 1 / (time.time() - last_time)
                print(f"[{current_time}] detection: {detection}, fps: {fps:.1f}, "
                      f"maxval: {detection_max_val:.0f}, maxloc: {detection_max_loc}, tx errors: {serial_tx_errors}")

            if detection != last_detection:
                if detection:
                    if debug:
                        print(f"[{current_time}] event: hand raised!")
                    serial_str = serial_msg_on_raise
                else:
                    if debug:
                        print(f"[{current_time}] event: hand lowered!")
                    serial_str = serial_msg_on_lowered

                # try to send via serial
                try:
                    if ser:
                        print(f"[{current_time}] sending message via serial: {serial_str}")
                        ser.write(serial_str)
                except SerialException as e:
                    serial_tx_errors = serial_tx_errors + 1
                    if e.errno == 13:  # write failed: [Errno 6] Device not configured
                        # retry once, may work or may fail; might want to re-open the serial
                        if ser:
                            print(f"[{current_time}] resending message via serial: {serial_str}")
                            ser.write(serial_str)
                    else:  # do not catch every exception
                        raise e

            # store for next iteration
            last_detection = detection


ms_teams_hand_detector.py
import argparse
import cv2
import numpy as np
import time
import mss
import serial
from datetime import datetime

if __name__ == "__main__":
    serial_tx_errors = 0

    # configure your messages to be sent on a
    # - first hand detected and
    # - last hand lowered events,
    # where the following symbols are used:
    # - 'P' to PLAY the hand raised sound,
    # - 'S' to move the SERVO motor to the hand raised position,
    # - 'A' to play the hand raised ANIMATION,
    # - 'p' to PLAY the hand lowered sound,
    # - 's' to move the SERVO motor to the hand lowered position,
    # - 'a' to play the hand lowered ANIMATION
    serial_msg_on_raise = b'PAS\x0d'
    serial_msg_on_lowered = b'pas\x0d'
    serial_msg_on_init = serial_msg_on_lowered

    debug = True  # set to false to have a very quiet mode

    parser = argparse.ArgumentParser(description='%(prog)s')

    parser.add_argument('-p', action="store", default=None, required=False,
                        dest='serport',
                        help="serial device port name (e.g. '/dev/cu.usbmodem14202' or 'COM1')")

    parser.add_argument('-t', action="store", default=4.2e6, required=False,
                        dest='threshold',
                        help="threshold value for detection of the hand symbol (e.g. '4.2e6')")

    parser.add_argument('-m', action="store", default=0, required=False,
                        dest='monitor_number',
                        help="monitor number (e.g. '0')")

    parser.add_argument('-r', action="store", default="the-hand.png", required=False,
                        dest='resource_path',
                        help="resource path to hand image (e.g. 'the-hand.png')")

    args = parser.parse_args()
    args.threshold = float(args.threshold)
    args.monitor_number = int(args.monitor_number)

    if debug:
        print(f"serport = {args.serport}")
        print(f"threshold = {args.threshold}")
        print(f"monitor_number = {args.monitor_number}")
        print(f"resource_path = {args.resource_path}")

    serial_baudrate = 115200

    ser = args.serport
    if ser:
        ser = serial.Serial(ser, serial_baudrate, timeout=3)

    if debug:
        print(f"ser = {ser}")

    # try to send initialization message via serial
    if ser:
        ser.write(serial_msg_on_init)

    template = cv2.imread(args.resource_path)
    template = cv2.cvtColor(template, cv2.COLOR_BGR2GRAY)
    template = cv2.Canny(template, 50, 200)
    w, h = template.shape[::-1]

    last_detection = False

    with mss.mss(display=':0') as sct:
        monitors = sct.monitors

        # list monitors
        if debug:
            mon_num = 0
            for mon in monitors:
                print(f"monitor #{mon_num} = {mon}")
                mon_num = mon_num + 1

        while True:
            last_time = time.time()

            img = sct.grab(monitors[args.monitor_number])
            img = np.array(img)
            gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
            edged = cv2.Canny(gray, 50, 200)

            result = cv2.matchTemplate(edged, template, cv2.TM_CCOEFF)
            (_, maxVal, _, maxLoc) = cv2.minMaxLoc(result)

            loc = np.where(result >= args.threshold)
            detection = len(loc[0]) > 0

            if debug:
                now = datetime.now()

                current_time = now.strftime("%H:%M:%S")

                fps = 1 / (time.time() - last_time)
                print(f"[{current_time}] detection: {detection}, fps: {fps:.1f}, tx errors: {serial_tx_errors}")

            if detection != last_detection:
                if detection:
                    if debug:
                        print(f"[{current_time}] event: hand raised!")
                    serial_str = serial_msg_on_raise
                else:
                    if debug:
                        print(f"[{current_time}] event: hand lowered!")
                    serial_str = serial_msg_on_lowered

                # try to send via serial
                try:
                    if ser:
                        print(f"[{current_time}] sending message via serial: {serial_str}")
                        ser.write(serial_str)
                except SerialException as e:
                    serial_tx_errors = serial_tx_errors + 1
                    if e.errno == 13:  # write failed: [Errno 6] Device not configured
                        # retry once, may work or may fail; might want to re-open the serial
                        if ser:
                            print(f"[{current_time}] resending message via serial: {serial_str}")
                            ser.write(serial_str)
                    else:  # do not catch every exception
                        raise e

            # store for next iteration
            last_detection = detection


gui.py
#!/usr/bin/python3
import tkinter as tk
import tkinter.ttk as ttk
import serial.tools.list_ports
import mss
import cv2
import time
import numpy as np
from datetime import datetime


lbl_det_rate_resetval = '-'
lbl_det_max_resetval = '-'
btn_det_startstop_start = 'Start'
btn_det_startstop_stop = 'Stop'
display_prefix = 'Display #'
combo_serial_dev_none = 'Do not use serial device'
raised_hand_detected_text = 'Raised hand detected'
raised_hand_not_detected_text = 'Raised hand not detected'
detection_not_started_text = 'Detection not started'
template_img_filename = 'the-hand.png'
serial_msg_on_raised_default = 'PAS\\r'
serial_msg_on_lowered_default = 'pas\\r'
processing_interval_msecs = 200  # also acts as rate limiter (max. 1000/processing_interval_msecs detections per second)
baudrates = sorted([9600, 19200, 38400, 57600, 115200, 230400], reverse=True)
baudrate_default = 115200
ser = None
selected_display_num = 0
detection_threshold = float('inf')
last_detection_result = None

last_time = time.time()
template_img = cv2.imread(template_img_filename)
template_img = cv2.cvtColor(template_img, cv2.COLOR_BGR2GRAY)
template_img = cv2.Canny(template_img, 50, 200)
template_img_w, template_img_h = template_img.shape[::-1]


def refresh_serial_dev_list():
    """rebuild list of serial devices"""
    combo_serial_dev['values'] = []
    for dev in serial.tools.list_ports.comports():
        combo_serial_dev['values'] = (*combo_serial_dev['values'], dev)
    combo_serial_dev['values'] = (combo_serial_dev_none, *combo_serial_dev['values'])
    combo_serial_dev['state'] = 'readonly'
    combo_serial_dev.current(0)


def refresh_display_list():
    """rebuild list of displays"""
    combo_display['values'] = []
    display_id = 0
    with mss.mss(display=':0') as sct:
        for display in sct.monitors:
            display_str = f"{display_prefix}{display_id} ({display['width']}x{display['height']} pixels)"
            combo_display['values'] = (*combo_display['values'], display_str)
            display_id += 1
    combo_display['state'] = 'readonly'
    combo_display.current(0)


def startstop_detection():
    """start or stop the detection"""
    global ser
    global selected_display_num
    global detection_threshold
    global last_detection_result
    global serial_msg_on_raised
    global serial_msg_on_lowered
    if btn_det_startstop['text'] == btn_det_startstop_start:
        # start
        btn_det_startstop['text'] = btn_det_startstop_stop
        ctrl_state = tk.DISABLED
        # get the display number
        selected_display = combo_display.get()
        selected_display = selected_display[len(display_prefix):]
        selected_display_num = int(selected_display.split()[0])
        # get the detection threshold
        detection_threshold = float(ent_det_thres.get())
        # extract serial device name and baudrate from dropdown menu and
        # create Serial object if valid
        serial_dev_selected = combo_serial_dev.get()
        if serial_dev_selected == combo_serial_dev_none:
            ser = None
        else:
            serial_dev_selected = serial_dev_selected.split(' ', 1)[0]
            serial_baudrate = combo_serial_speed.get()
            ser = serial.Serial(serial_dev_selected, serial_baudrate, timeout=3)
        # save messages to be sent via serial
        serial_msg_on_raised = ent_msg_raised.get().replace('\\r', '\r').replace('\\n', '\n').encode()
        serial_msg_on_lowered = ent_msg_lowered.get().replace('\\r', '\r').replace('\\n', '\n').encode()
    else:
        # stop: reset some of the GUI elements
        btn_det_startstop['text'] = btn_det_startstop_start
        lbl_det_rate['text'] = lbl_det_rate_resetval
        lbl_det_max['text'] = lbl_det_max_resetval
        lbl_det_status['text'] = detection_not_started_text
        last_detection_result = None
        ctrl_state = tk.NORMAL
        # close serial device if Serial object is valid
        if ser:
            ser.close()
    # change state of GUI control elements (activate/deactivate)
    ent_det_thres['state'] = ctrl_state
    combo_display['state'] = ctrl_state
    btn_display_refresh['state'] = ctrl_state
    combo_serial_dev['state'] = ctrl_state
    btn_serial_dev_refresh['state'] = ctrl_state
    combo_serial_speed['state'] = ctrl_state
    ent_msg_raised['state'] = ctrl_state
    ent_msg_lowered['state'] = ctrl_state


def worker():
    """perform the detection (core)"""
    global last_time
    global template_img
    global selected_display_num
    global detection_threshold
    global last_detection_result
    global ser
    global serial_msg_on_raised
    global serial_msg_on_lowered
    # do the real work if detection is activated
    if btn_det_startstop['text'] == btn_det_startstop_stop:
        sct = mss.mss(display=':0')
        img = sct.grab(sct.monitors[selected_display_num])
        img = np.array(img)
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        edged = cv2.Canny(gray, 50, 200)

        result = cv2.matchTemplate(edged, template_img, cv2.TM_CCOEFF)
        (_, detection_max_val, _, detection_max_loc) = cv2.minMaxLoc(result)
        lbl_det_max['text'] = f"{detection_max_val:.2e} @ pos. {detection_max_loc}"

        loc = np.where(result >= detection_threshold)
        detection_result = len(loc[0]) > 0

        if detection_result:
            lbl_det_status['text'] = raised_hand_detected_text
        else:
            lbl_det_status['text'] = raised_hand_not_detected_text
        # edge detection
        if last_detection_result != detection_result:
            if detection_result:
                serial_str = serial_msg_on_raised
            else:
                serial_str = serial_msg_on_lowered

            # try to send via serial
            try:
                if ser:
                    ser.write(serial_str)
            except serial.SerialException as e:
                if e.errno == 13:  # write failed: [Errno 6] Device not configured
                    # retry once, may work or may fail; might want to re-open the serial
                    if ser:
                        ser.write(serial_str)
                else:  # do not catch every exception
                    raise e
        # store current detection result for next run (for edge detection)
        last_detection_result = detection_result

        # calculate detection rate
        this_time = time.time()
        detections_per_sec = 1 / (this_time - last_time)
        lbl_det_rate['text'] = f"{detections_per_sec:.2f}"
        last_time = this_time

    window.after(processing_interval_msecs, worker)  # reschedule


# build up the GUI
# (first the window, then all elements for every row of the grid)
window = tk.Tk()
window.title('Microsoft Teams Hand Detector')

cur_row = 0

lbl_det_status_descr = tk.Label(text='Detection status')
lbl_det_status_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

lbl_det_status = tk.Label(text=detection_not_started_text)
lbl_det_status.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')

btn_det_startstop = tk.Button(text=btn_det_startstop_start, command=startstop_detection)
btn_det_startstop.grid(row=cur_row, column=2, padx='5', pady='5', sticky='w')


cur_row += 1

lbl_det_rate_descr = tk.Label(text='Detection refresh rate (1/s)')
lbl_det_rate_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

lbl_det_rate = tk.Label(text=lbl_det_rate_resetval)
lbl_det_rate.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')


cur_row += 1

lbl_det_max_descr = tk.Label(text='Detection max. value')
lbl_det_max_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

lbl_det_max = tk.Label(text=lbl_det_max_resetval)
lbl_det_max.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')


cur_row += 1

lbl_det_thres_descr = tk.Label(text='Detection threshold')
lbl_det_thres_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

ent_det_thres = tk.Entry()
ent_det_thres.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')
ent_det_thres.insert(0, '4.2e6')


cur_row += 1

lbl_display_descr = tk.Label(text='Display')
lbl_display_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

combo_display = ttk.Combobox()
combo_display.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')

btn_display_refresh = tk.Button(text='Refresh', command=refresh_display_list)
btn_display_refresh.grid(row=cur_row, column=2, padx='5', pady='5', sticky='w')


cur_row += 1

lbl_serial_dev_descr = tk.Label(text='Serial device')
lbl_serial_dev_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

combo_serial_dev = ttk.Combobox()
combo_serial_dev.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')

btn_serial_dev_refresh = tk.Button(text='Refresh', command=refresh_serial_dev_list)
btn_serial_dev_refresh.grid(row=cur_row, column=2, padx='5', pady='5', sticky='w')


cur_row += 1

lbl_serial_speed_descr = tk.Label(text='Serial baudrate')
lbl_serial_speed_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

combo_serial_speed = ttk.Combobox()
combo_serial_speed.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')
for br in baudrates:
    combo_serial_speed['values'] = (*combo_serial_speed['values'], (br))
combo_serial_speed.current(baudrates.index(baudrate_default))

cur_row += 1

lbl_msg_raised_descr = tk.Label(text='Msg. on hand raised event')
lbl_msg_raised_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

ent_msg_raised = tk.Entry()
ent_msg_raised.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')
ent_msg_raised.insert(0, serial_msg_on_raised_default)


cur_row += 1

lbl_msg_lowered_descr = tk.Label(text='Msg. on hand lowered event')
lbl_msg_lowered_descr.grid(row=cur_row, column=0, padx='5', pady='5', sticky='w')

ent_msg_lowered = tk.Entry()
ent_msg_lowered.grid(row=cur_row, column=1, padx='5', pady='5', sticky='w')
ent_msg_lowered.insert(0, serial_msg_on_lowered_default)


# call functions to fill lists initially
refresh_serial_dev_list()
refresh_display_list()

# kick off the GUI
window.after(processing_interval_msecs, worker)
window.mainloop()
